[ZOOKEEPER-1525] Plumb ZooKeeperServer object into auth plugins

Randgalt opened the pull request:
Based on patch work from https://issues.apache.org/jira/browse/ZOOKEEPER-1525

Created ServerAuthenticationProvider which has a method to accept the ZooKeeper
server so that auth can be done using values in the ZK database. As this is a new
interface, existing implementations aren't affected helping backward compatibility

eribeiro commented on src/java/main/org/apache/zookeeper/server/auth/KeyAuthenticationProvider.java:
```diff
@@ -0,0 +1,146 @@
+            if (keyVal!=0 && ((authVal % keyVal) != 0)) {
+              return false;
+            }
+        } catch (NumberFormatException nfe) {
```
nit: ZK support Java 7, so we could rewrite this line as below (i.e., multi-catch), right?
```
} catch (NumberFormatException|UnsupportedEncodingException ex) {
      return false;
}
```

eribeiro replied:
As ZK supports Java7 we could use multi-catch, right? Like:

```
} catch (NumberFormatException|UnsupportedEncodingException ex) {
     return false;
}
```

eribeiro commented on src/java/test/org/apache/zookeeper/test/KeyAuthClientTest.java:
```diff
@@ -0,0 +1,141 @@
+        try {
+            zk.setData("/abc", "testData2".getBytes(), -1);
+            Assert.fail("Should not be able to set data");
+        } catch (KeeperException e) {
```
same as above

eribeiro replied:
Maybe insert a comment like: // test passed or // ignore

Randgalt replied:
I used KeyAuthClientTest.java unchanging from the original PR. I'll have a pass at it to clean things up.

eribeiro commented on src/java/test/org/apache/zookeeper/test/KeyAuthClientTest.java:
```diff
@@ -0,0 +1,141 @@
+        try {
+            zk.getData("/abc", false, null);
+            Assert.fail("Should not be able to get data");
+        } catch (KeeperException e) {
```
nit: insert a comment like '// test passed' in the catch block?

eribeiro commented on src/java/test/org/apache/zookeeper/test/KeyAuthClientTest.java:
```diff
@@ -0,0 +1,141 @@
+
+public class KeyAuthClientTest extends ClientBase {
+    private static final Logger LOG = LoggerFactory.getLogger(KeyAuthClientTest.class);
+    static final ArrayList<ACL> KeyACLs = new ArrayList<ACL>(
```
``KeyACLS`` seems to be never used.

eribeiro replied:
``KeyACLS`` seems to be never used...

eribeiro commented on src/java/test/org/apache/zookeeper/test/KeyAuthClientTest.java:
```diff
@@ -0,0 +1,141 @@
+          LOG.debug("  "+acl.toString());
+        }
+      } catch (Exception e) {
+          LOG.debug("  EXCEPTION THROWN");
```
Maybe add the reason as:

```
LOG.debug("   EXCEPTION THROWN ", e);
```

fpj commented on src/java/main/org/apache/zookeeper/server/auth/ProviderRegistry.java:
```diff
@@ -61,17 +64,17 @@ public static void initialize() {
     }
 
-    public static AuthenticationProvider getProvider(String scheme) {
+    public static ServerAuthenticationProvider getProvider(ZooKeeperServer zks, String scheme) {
```
What if we make the signature `Object obj, String scheme` or maybe a variable number of arguments `Object..., String scheme` to enable future changes to pass other instances through the `getProvider` method?

Randgalt replied:
Loose typing leads to errors in my experience. Maybe we could pass a map of meta data?

fpj replied:
Yeah, I agree that loose typing can lead to errors. My concern is that we haven't needed to pass a `ZooKeeperServer` object for the previous `AuthenticationProvider` implementations, and with the current changes all calls to `getProvider` and `initialize` get such an instance. A metadata map is possibly a better way to go.

fpj replied:
If we pass it to `handleAuthentication` and `matchesOp` as you suggested is slightly better, but it still has the problem that most `AuthenticationProvider` implementations do not use an instance of `ZooKeeperServer`.

Randgalt replied:
We can un-deprecate the existing `AuthenticationProvider` and let people choose which API they want.

fpj commented on src/java/main/org/apache/zookeeper/server/auth/ProviderRegistry.java:
```diff
@@ -33,7 +33,7 @@
         new HashMap<String, AuthenticationProvider>();
 
-    public static void initialize() {
+    public static void initialize(ZooKeeperServer zks) {
```
I don't quite like the idea of passing a `ZooKeeperServer` instance here, it is too specific. See suggestion below.

fpj commented on src/java/main/org/apache/zookeeper/server/auth/ProviderRegistry.java:
```diff
@@ -61,17 +64,17 @@ public static void initialize() {
-            initialize();
-        return authenticationProviders.get(scheme);
+            initialize(zks);
+        return WrappedAuthenticationProvider.wrap(authenticationProviders.get(scheme));
```
I'm not sure I like this wrapping. In this PR, we are only calling `setZooKeeperServer` in the `initialize` method above, so I'm wondering whether `setZooKeeperServer` really needs to be part of the interface and whether we can make `setZooKeeperServer` just a method of the `ServerAuthenticationProvider` implementation, which we can still set during initialization. 

If we do something like this, then we don't need to deprecate `AuthenticationProvider`. What do you think?

Randgalt replied:
The wrapping is also needed for the new matchesOp method with replaces matches. The Wrapper forwards to the old method for old implementations that implement `AuthenticationProvider`. This achieves backward compatibility. Can you think of another way?

Randgalt replied:
Another possibility - pass the ZooKeeperServer as an argument to `handleAuthentication()` and `matchesOp()`.

fpj replied:
Let me take a step back. It feels that the `ZooKeeperServer` object is really part of `ServerAuthenticationProvider` but not to the generic interface. Specifically, it feels like we need some interface that I'm going to call A, and both `AuthenticationProvider` and  `ServerAuthenticationProvider` specialize A.

The trouble of going down this path I'm suggesting is that we will need to make in the code to know whether we are dealing with one type or the other to determine which methods to call. For example, `matches` belongs in `AuthenticationProvider` while `matchesOp` belongs in `ServerAuthenticationProvider`.

Randgalt replied:
Differentiating between `ServerAuthenticationProvider` and `AuthenticationProvider` is where the beauty of `WrappedAuthenticationProvider` comes in. `WrappedAuthenticationProvider` does this differentiating. The rest of the code only deals with one API and `WrappedAuthenticationProvider` mitigates the two.

fpj replied:
Yeah, I get that `WrappedAuthenticationProvider` is essentially transforming an `AuthenticationProvider` implementation into a `ServerAuthenticationProvider`. My concern is that it feels like the wrong abstraction to me because the implementations we currently have are independent of `ZooKeeperServer` instance. 

The fact that `ServerAuthenticationProvider` depends on a `ZooKeeperServer` instance is specific to the class of providers the interface is representing, but not all. The abstraction doesn't quite work in my head when I think that we are trying to make all `AuthenticationProvider` implementations an instance of `ServerAuthenticationProvider` via the wrapper.

In general, I get how it works, I'm just not super convinced that we are reflecting the right abstraction in the code.

Randgalt replied:
OK - seems like a good set of abstractions to me. Let me know how you'd like it to change.

Randgalt commented on src/java/main/org/apache/zookeeper/server/ServerCnxn.java:
```diff
@@ -51,7 +47,7 @@
     private static final Logger LOG = LoggerFactory.getLogger(ServerCnxn.class);
     
-    protected ArrayList<Id> authInfo = new ArrayList<Id>();
+    private Set<Id> authInfo = Collections.newSetFromMap(new ConcurrentHashMap<Id, Boolean>());
```
@fpj this is a more aggressive change but one for the better. 1) there's no good reason to have this field be protected; 2) it was defined as a List but used as a Set - so correct that; 3) make it concurrent/thread safe

fpj commented:
@Randgalt Here are my current thoughts on the abstraction discussion. The main issue is that `AuthenticationProvider` implementations take no constructor parameter, and we are trying to introduce one that requires an input parameter. My two main goals are:

1- To have `getProvider` to return a base class rather than wrapping
2- To have a more general mechanism to pass parameters in the case future changes require other parameters

I suggest we create a base class that we call `BaseAuthenticationProvider`. `BaseAuthenticationProvider` is the current `AuthenticationProvider` without the `matches` call. Both `ServerAuthenticationProvider` and `AuthenticationProvider` extend `BaseAuthenticationProvider`. The `getProvider` call should also get a set of generic parameters that it passes to the provider implementations. Implementations may choose to ignore or use any of the parameters. 

You suggest passing a set of generic parameters with a metadata map. As I understand it, it would be a map from a `BaseAuthenticationProvider` class to a set of objects or some form of container that is provider specific.

Let me know if this makes sense to you.

Randgalt commented:
@fpj I don't see how this can be done without breaking existing code. If we introduce a new base class then all the current `AuthenticationProvider`s have to change. Or, we'd have to do runtime `instanceof` checks or something to differentiate which is a mistake and error prone. How do you propose to mitigate this?

fpj commented:
@Randgalt actually, I'm thinking that if we get to pass a set of parameters generically to an `AuthenticationProvider`, then we don't even need new provider interfaces. If I'm not mistaken, all we need to do is to pass the metadata we have discussed as a constructor parameter and let the `AuthenticationProvider` implementation deal with the parameters internally. Does it work?

Randgalt commented:
@fpj Unfortunately not. All current versions directly implement `AuthenticationProvider`. We're not at Java 8 so we can't add default methods. Any change at all to AuthenticationProvider will break existing code.

fpj commented:
yeah, I realized that after sending the comment, thinking some more...

Randgalt commented:
@fpj that's why I added the Wrapper class. I don't see any way around it. It solves the problem nicely.

fpj commented:
My main concern here is with the generality of the approach. I think that when we did the `AuthenticationProvider` stuff originally, we were thinking that we wouldn't need parameters for the constructor of provider implementations or new parameters to the `matches` method. 

I'm with you that your wrapper proposal fixes the issue for `ServerAuthenticationProvider`, but I'm concerned that in the future we might realize that we need some other parameters for a different class of `AuthenticationProvider` implementations, and we shove them into the `getProvider` signature again. We will also need some more wrapping because we can only return one type in `getProvider`, we make a single call to `matches` in `PrepRequestProcessor`, and we don't want to do ìnstanceof` all the time.

At a high level, we need two things:
1. A signature for `matches` that serves both implementations of `AuthenticationProvider` and implementations of `ServerAuthenticationProvider`. 
1. A generic way to pass parameters to build `AuthenticationProvider` implementations.

Assuming that the above makes sense, one possibility is that we change the signature of `matches` to take a generic set of parameters. That's not great either because once you generalize, you're opening up for errors. So we are left with pushing the problem for later, if we have ever the need of adding another provider interface, or generalizing and opening up for errors.

Randgalt commented:
@fpj perfection is the enemy of the good right? The current implementation has worked well for a long time. I'd rather see us pick a reasonable set of arguments now that try to have something that will work for a possible future. I think removing the ZookeeperServer from the getProvider() is actually a better solution that what we have here now. I'll work on and push that. Let's discuss after I've refined this a bit.

Randgalt commented:
@fpj Let me know what you think of this: I reworked ServerAuthenticationProvider so that it's an abstract class. This provides a clearer distinction between the two AuthenticationProviders and allows subclasses of ServerAuthenticationProvider to not have to implement the older matches() and handleAuthentication(). Additionally, removed the method to set the ZooKeeperServer in favor of passing it as a parameter as needed.

fpj commented:
@Randgalt It does look better, thanks for reworking it. Let me make another pass pro detail, but I think the general approach looks good.

Randgalt commented:
@fpj Any more thoughts on this Flavio?

hanm commented:
(Not particularly related to the logic of this PR) Might be better to do a git squash to consolidate 16 commits into a single commit to keep the commit history clean - I've seen most projects operate that way.

fpj commented on src/java/test/org/apache/zookeeper/test/KeyAuthClientTest.java:
```diff
@@ -0,0 +1,131 @@
+        LOG.debug("KeyAuthenticationProvider Creating Test Node:"+path+".\n");
+        zk.create(path, null, Ids.CREATOR_ALL_ACL, CreateMode.PERSISTENT);
+        List<ACL> acls = zk.getACL(path, null);
+        LOG.debug("Node: "+path+" Test:"+testName+" ACLs:");
```
space around symbols, please?

Randgalt replied:
I didn't write this code - only copied it. But, I'll reformat the file to style.

fpj commented on src/java/test/org/apache/zookeeper/test/KeyAuthClientTest.java:
```diff
@@ -0,0 +1,131 @@
+        List<ACL> acls = zk.getACL(path, null);
+        LOG.debug("Node: "+path+" Test:"+testName+" ACLs:");
+        for (ACL acl : acls) {
+          LOG.debug("  "+acl.toString());
```
Same here, space around symbols, please.

fpj commented on src/java/test/org/apache/zookeeper/test/KeyAuthClientTest.java:
```diff
@@ -0,0 +1,131 @@
+
+    @Test
+    public void testAuth() throws Exception {
+        // NOTE: the tests need to run in-order, and older versions of
```
I think it is fine if you have a sequence of steps in your test case, but I really see this as a single test case with multiple steps. I suggest we remove the `test` prefix from the methods to make clear that they aren't independent test cases.

fpj commented on src/java/test/org/apache/zookeeper/test/KeyAuthClientTest.java:
```diff
@@ -0,0 +1,131 @@
+            zk.setData("/abc", "testData1".getBytes(), -1);
+            zk.create("/key", null, Ids.CREATOR_ALL_ACL, CreateMode.PERSISTENT);
+            zk.setData("/key", "5".getBytes(), -1);
+            Thread.sleep(1000);
```
I'm not sure why we need this `sleep` here.

Randgalt replied:
Me neither - I didn't write this code.

eribeiro replied:
Only thing I can suppose (and **please** correct me if I am saying something **utterly stupid**) is that this ``sleep`` was inserted in the hopes of giving some time to ZK commit the changes. Naive but yet (shrug)... 

Well, I guess we can either remove it or replace it with a ``zk.getData("/key");``, right?

Randgalt replied:
These types of sleeps are all over the ZK tests (and Curator for that matter).

eribeiro replied:
So, a question for a future ticket (IMHO). :cactus:

fpj commented on src/java/main/org/apache/zookeeper/server/auth/ProviderRegistry.java:
```diff
@@ -31,7 +31,15 @@
+        new HashMap<>();
+
+    //VisibleForTesting
+    public static void reset() {
```
Is this used anywhere?

fpj commented on src/java/main/org/apache/zookeeper/server/auth/ServerAuthenticationProvider.java:
```diff
@@ -0,0 +1,79 @@
+     *                for set ACL operations, the list of ACLs being set. Otherwise null.
+     * @return true if the arguments can be matched by the expression.
+     */
+    public abstract boolean matches(ZooKeeperServer zks, ServerCnxn cnxn, String path, String id, String aclExpr, int perm, List<ACL> setAcls);
```
Can we break this line, please?

fpj commented on src/java/main/org/apache/zookeeper/server/auth/WrappedAuthenticationProvider.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    /**
+     * {@inheritDoc}
```
I've never used this before and from the name it implies that it will copy the doc blurb from the super class when running the javadoc tool. Is this correct? Do we have to do it for this method only or for others as well?

Randgalt replied:
Yes, you are correct. 

>Do we have to do it for this method only or for others as well?
Well, it can be done anywhere. I only need it here.

fpj commented:
@Randgalt My comments are pretty minor, but it will give it a chance to QA to test this PR.

Randgalt commented on src/docs/src/documentation/content/xdocs/zookeeperProgrammers.xml:
```diff
@@ -1193,6 +1193,55 @@ authProvider.2=com.f.MyAuth2
+    </para>
+
+    <programlisting>
+public abstract class ServerAuthenticationProvider implements AuthenticationProvider {
```
I'm copying the other parts of the zookeeperProgrammers.xml here. I built the docs and it looks good.

fpj replied:
Ok.

Randgalt commented on src/java/main/org/apache/zookeeper/server/PrepRequestProcessor.java:
```diff
@@ -543,7 +546,7 @@ protected void pRequest2Txn(int type, long zxid, Request request,
                 
                 nodeRecord = getRecordForPath(ZooDefs.CONFIG_NODE);               
-                checkACL(zks, nodeRecord.acl, ZooDefs.Perms.WRITE, request.authInfo);                  
+                checkACL(zks, request.cnxn, nodeRecord.acl, ZooDefs.Perms.WRITE, request.authInfo, "/", null);
```
Something needs to be passed. I guess I can pass null. What do you think?

fpj replied:
`null` tells me that the method is not expecting a value from the caller in that position, which is better than an arbitrary value. I think it is fine to change it to `null`.

Randgalt replied:
done

Randgalt commented:
@fpj I've addressed and pushed changes for all your comments.

Randgalt commented:
@fpj any updates on this?

fpj commented:
@Randgalt I'm sorry that QA isn't working for pull requests yet, could please create a diff patch and upload to the jira so that we can get QA to run on it?

Randgalt commented:
latest patch added

lvfangmin commented on src/java/main/org/apache/zookeeper/server/auth/KeyAuthenticationProvider.java:
```diff
@@ -0,0 +1,139 @@
+        return "key";
+    }
+
+    public byte[] getKey(ZooKeeperServer zks) {
```
Do we need this to be public? I only saw it's being called in handleAuthentication.

lvfangmin commented:
Generally looks good to me, but I have the same concern as @fpj, maybe in the future the newly added auth needs additional parameters, then we have to add another wrapper with new interface to keep the old ones backward compatible. I also agree a general metadata map is error-prone, but I think we can have a compromise solution between explicit parameters and general metadata map: 

Create a class (e.g. AuthData) with the explicit parameters(zks, cnxn, path, perm, setAcls) in it, which has the get and set method, and use this class in the newly added ServerAuthenticationProvider interface. In the future if we need additional parameter, we only need to change this class, no interface and backward compatible issues will be involved. What do you think @Randgalt @fpj?

Randgalt commented:
OK - I've put the args into container classes. Let me know what you think.

fpj commented:
@Randgalt thanks for the recent changes, they look fine for me, except that we didn't get QA to run on it.

@lvfangmin there are a few levels of indirection here, which doesn't make it look very pretty, but I don't see a much better way of doing it. how do you see the recent changes, do they look goo to you?
